src/pyclass.rs

```diff
@@ -154,122 +190,197 @@ unsafe fn create_type_object_impl(
+            // Safety: This is the correct slot type for Py_sq_item
+            unsafe {
+                self.push_slot(
+                    ffi::Py_sq_item,
+                    get_sequence_item_from_mapping as *mut c_void,
+                )
+            }
         }
 
-        if has_setitem {
-            push_slot(
-                &mut slots,
-                ffi::Py_sq_ass_item,
-                assign_sequence_item_from_mapping as _,
-            );
+        if !self.is_mapping && self.has_setitem {
+            // Safety: This is the correct slot type for Py_sq_ass_item
+            unsafe {
+                self.push_slot(
+                    ffi::Py_sq_ass_item,
+                    assign_sequence_item_from_mapping as *mut c_void,
+                )
+            }
         }
     }
 
-    if !has_new {
-        push_slot(&mut slots, ffi::Py_tp_new, no_constructor_defined as _);
+    fn set_is_basetype(mut self, is_basetype: bool) -> Self {
+        if is_basetype {
+            self.class_flags |= ffi::Py_TPFLAGS_BASETYPE;
+        }
+        self
     }
 
-    if has_clear && !has_traverse {
-        return Err(PyTypeError::new_err(format!(
-            "`#[pyclass]` {} implements __clear__ without __traverse__",
-            name
-        )));
+    fn set_is_mapping(mut self, is_mapping: bool) -> Self {
+        self.is_mapping = is_mapping;
+        self
     }
 
-    // Add empty sentinel at the end
-    push_slot(&mut slots, 0, ptr::null_mut());
-
-    let mut spec = ffi::PyType_Spec {
-        name: py_class_qualified_name(module_name, name)?,
-        basicsize: basicsize as c_int,
-        itemsize: 0,
-        flags: py_class_flags(has_traverse, is_basetype),
-        slots: slots.as_mut_ptr(),
-    };
-
-    let type_object = ffi::PyType_FromSpec(&mut spec);
-    if type_object.is_null() {
-        Err(PyErr::fetch(py))
-    } else {
-        tp_init_additional(
-            type_object as _,
-            tp_doc,
-            #[cfg(all(not(Py_3_9), not(Py_LIMITED_API)))]
-            &buffer_procs,
-            #[cfg(not(Py_3_9))]
-            dict_offset,
-            #[cfg(not(Py_3_9))]
-            weaklist_offset,
-        );
-        Ok(type_object as _)
+    /// # Safety
+    /// All slots in the PyClassItemsIter should be correct
+    unsafe fn with_class_items(mut self, iter: PyClassItemsIter) -> Self {
+        for items in iter {
+            for slot in items.slots {
+                self.push_slot(slot.slot, slot.pfunc);
+            }
+            for method in items.methods {
+                self.add_pymethod_def(method);
+            }
+        }
+        self
     }
-}
 
-#[cold]
-fn type_object_creation_failed(py: Python<'_>, e: PyErr, name: &'static str) -> ! {
-    e.print(py);
-    panic!("An error occurred while initializing class {}", name)
-}
+    fn with_type_doc(mut self, type_doc: &'static str) -> Self {
+        if let Some(doc) = py_class_doc(type_doc) {
```

Comment:
Because you're taking a `&'static str`, this means `py_class_doc` shouldn't have to leak that CString.

Reply:
Is that true? If the `str` isn't null-terminated the copy will need to be made.

Reply:
It usually will be null terminated. if it's not, it will still need to copy+leak that cstring yes.

Reply:
> It usually will be null terminated.

So could we enforce this invariant by taking `&'static CStr`?

Reply:
I've wanted to do this in the past but I think const-constructed `CStr` isn't possible on MSRV. I guess could leak earlier for older compilers.

Reply:
If we enforce the `type_doc` is always a null-terminated `&'static str` I believe we could cast it directly into a pointer without leaking the memory. We could easily check that invariant at runtime to keep this function safe.

Reply:
For now, I'd like this PR to keep the behavior exactly the same as before, so I won't make this change, but I will in a followup PR.